File: android/Xtrade/app/src/main/java/nz/co/fendustries/xtrade/fragments/AllExchangeRatesFragment.java
package nz.co.fendustries.xtrade.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import javax.inject.Inject;

import dagger.Lazy;
import nz.co.fendustries.xtrade.R;
import nz.co.fendustries.xtrade.XtradeApplication;
import nz.co.fendustries.xtrade.activities.ExchangeRateDetailsActivity;
import nz.co.fendustries.xtrade.activities.HomeActivity;
import nz.co.fendustries.xtrade.adapters.RatesRecyclerAdapter;
import nz.co.fendustries.xtrade.helpers.AndroidConstants;
import nz.co.fendustries.xtrade.interfaces.AllExchangeRatesViewContract;
import nz.co.fendustries.xtrade.interfaces.RecyclerViewItemTapCallback;
import nz.co.fendustries.xtrade.presenters.AllExchangeRatesPresenter;

/**
 * Created by joshuafenemore on 21/10/16.
 */
public class AllExchangeRatesFragment extends Fragment implements AllExchangeRatesViewContract
{
    private TextView noRatesTextView;
    private SwipeRefreshLayout swipeRefreshLayout;
    private RecyclerView ratesRecyclerView;
    private RecyclerView.LayoutManager ratesRecylerViewLayoutManager;
    private RatesRecyclerAdapter ratesRecyclerAdapter;

    @Inject
    Lazy<AllExchangeRatesPresenter> allExchangeRatesPresenterLazy;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        super.onCreateView(inflater, container, savedInstanceState);
        ((XtradeApplication)this.getActivity().getApplication()).getXtradeComponent().inject(this);

        View view = inflater.inflate(R.layout.fragment_all_exchange_rates, container, false);

        this.noRatesTextView = (TextView) view.findViewById(R.id.noRatesTextView);
        this.swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.swipeRefreshLayout);
        this.ratesRecyclerView = (RecyclerView) view.findViewById(R.id.ratesRecyclerView);
        this.ratesRecylerViewLayoutManager = new LinearLayoutManager(this.getActivity());
        this.ratesRecyclerView.setLayoutManager(this.ratesRecylerViewLayoutManager);

        return view;
    }

    @Override
    public void onResume()
{
    super.onResume();

    ((HomeActivity)this.getActivity()).setActionBarTitle(this.getString(R.string.allRatesLabel));

    this.swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener()
    {
        @Override
        public void onRefresh()
        {
            allExchangeRatesPresenterLazy.get().refreshRate();
        }
    });

    this.allExchangeRatesPresenterLazy.get().setAllExchangeRatesViewContract(this);
}

    @Override
    public void onPause()
    {
        super.onPause();

        this.allExchangeRatesPresenterLazy.get().setAllExchangeRatesViewContract(null);

        this.swipeRefreshLayout.setRefreshing(false);
        this.swipeRefreshLayout.setOnRefreshListener(null);
        this.swipeRefreshLayout.destroyDrawingCache();
        this.swipeRefreshLayout.clearAnimation();
    }

    @Override
    public void showRatesView()
    {
        this.getActivity().runOnUiThread(new Runnable()
        {
            @Override
            public void run()
            {
                noRatesTextView.setVisibility(View.GONE);
                ratesRecyclerView.setVisibility(View.VISIBLE);

                ratesRecyclerAdapter = new RatesRecyclerAdapter((AppCompatActivity) getActivity(), allExchangeRatesPresenterLazy.get().getAllRates(), new RecyclerViewItemTapCallback()
                {
                    @Override
                    public void onItemTap(int position)
                    {
                        Intent detailsIntent = new Intent(getActivity(), ExchangeRateDetailsActivity.class);
                        detailsIntent.putExtra(AndroidConstants.SELECTED_RATE_CODE, allExchangeRatesPresenterLazy.get().getAllRates().get(position).getCurrencyCode());
                        startActivity(detailsIntent);
                    }
                });

                ratesRecyclerView.setAdapter(ratesRecyclerAdapter);
            }
        });
    }

    @Override
    public void showRefreshingView(boolean isRefreshing)
    {
        this.swipeRefreshLayout.setRefreshing(isRefreshing);
    }

    @Override
    public String getRefreshSuccessMessage()
    {
        return this.isVisible() ? this.getString(R.string.rates_refreshed) : "";
    }

    @Override
    public String getRefreshFailedMessage()
    {
        return this.isVisible() ? this.getString(R.string.rates_refresh_fail) : "";
    }

    @Override
    public void showRefreshMessage(String message)
    {
        if (this.isVisible())
        {
            Snackbar.make(this.ratesRecyclerView, message, Snackbar.LENGTH_LONG).show();
        }
    }

    @Override
    public void showNoRatesView()
    {
        this.getActivity().runOnUiThread(new Runnable()
        {
            @Override
            public void run()
            {
                ratesRecyclerView.setVisibility(View.GONE);
                noRatesTextView.setVisibility(View.VISIBLE);
            }
        });
    }
}

File: xamarin/Xtrade/Xtrade.AndroidApp/fragments/AllExchangeRatesFragment.cs
namespace Xtrade.AndroidApp.Fragments
{
    using System;
    using Activities;
    using Adapters;
    using Android.Content;
    using Android.OS;
    using Android.Support.Design.Widget;
    using Android.Support.V4.Widget;
    using Android.Support.V7.App;
    using Android.Support.V7.Widget;
    using Android.Views;
    using Android.Widget;
    using Shared.Interfaces.ViewModels;

    public class AllExchangeRatesFragment : BaseFragment<IAllRatesViewModel>
    {
        private TextView _noRatesTextView;
        private SwipeRefreshLayout _swipeRefreshLayout;
        private RecyclerView _ratesRecyclerView;
        private RecyclerView.LayoutManager _ratesRecylerViewLayoutManager;
        private RatesRecyclerAdapter _ratesRecyclerAdapter;

        public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
        {
            base.OnCreateView(inflater, container, savedInstanceState);

            View view = inflater.Inflate(Resource.Layout.fragmnety_all_exchange_rates, container, false);

            this._noRatesTextView = view.FindViewById<TextView>(Resource.Id.noRatesTextView);
            this._swipeRefreshLayout = view.FindViewById<SwipeRefreshLayout>(Resource.Id.swipeRefreshLayout);
            this._ratesRecyclerView = view.FindViewById<RecyclerView>(Resource.Id.ratesRecyclerView);
            this._ratesRecylerViewLayoutManager = new LinearLayoutManager(this.Activity);

            this.Activity.ActionBar.SetTitle(Resource.String.allRatesLabel);

            return view;
        }

        public override void OnResume()
        {
            base.OnResume();

            this.ViewModel.OnViewModelDataChanged += this.ViewModelDataChanged;
            this.ViewModel.OnRefreshError += this.ViewModelRefreshError;
            this.ViewModel.OnRefreshSuccess += this.ViewModelRefreshSuccess;

            this._swipeRefreshLayout.Refresh += this.SwipeRefreshLayoutOnRefresh;

            this._swipeRefreshLayout.Refreshing = true;
            this.ViewModel.LoadData();
        }

        public override void OnPause()
        {
            base.OnPause();

            this.ViewModel.OnViewModelDataChanged -= this.ViewModelDataChanged;
            this.ViewModel.OnRefreshError -= this.ViewModelRefreshError;
            this.ViewModel.OnRefreshSuccess -= this.ViewModelRefreshSuccess;

            this._swipeRefreshLayout.Refreshing = false;
            this._swipeRefreshLayout.Refresh -= this.SwipeRefreshLayoutOnRefresh;
            this._swipeRefreshLayout.DestroyDrawingCache();
            this._swipeRefreshLayout.ClearAnimation();
        }

        private void UpdateViews()
        {
            this._swipeRefreshLayout.Refreshing = this.ViewModel.IsDataRefreshing;

            if (this.ViewModel.AllRates == null || this.ViewModel.AllRates.Count == 0)
            {
                this._noRatesTextView.Visibility = ViewStates.Visible;
                this._ratesRecyclerView.Visibility = ViewStates.Gone;
            }
            else
            {
                this._noRatesTextView.Visibility = ViewStates.Gone;
                this._ratesRecyclerView.Visibility = ViewStates.Visible;

                if (this._ratesRecyclerAdapter == null)
                {
                    this._ratesRecyclerAdapter = new RatesRecyclerAdapter((AppCompatActivity)this.Activity, this.ViewModel.AllRates, i =>
                    {
                        Intent detailsIntent = new Intent(this.Activity, typeof (ExchangeRateDetailsActivity));
                        detailsIntent.PutExtra(Helpers.AndroidConstants.SelectedRateCode, this.ViewModel.AllRates[i].CurrencyCode);
                        this.StartActivity(detailsIntent);
                    });

                    this._ratesRecyclerView.SetAdapter(this._ratesRecyclerAdapter);
                    this._ratesRecyclerView.SetLayoutManager(this._ratesRecylerViewLayoutManager);
                }
                else
                {
                    this._ratesRecyclerAdapter.UpdateDataSet(this.ViewModel.AllRates);
                }
            }
        }

        private void SwipeRefreshLayoutOnRefresh(object sender, EventArgs eventArgs)
        {
            this._swipeRefreshLayout.Refreshing = true;
            this.ViewModel.RefreshRates();
        }

        private void ViewModelDataChanged(object sender, EventArgs eventArgs)
        {
            this.Activity.RunOnUiThread(() =>
            {
                if (this.IsVisible)
                {
                    this.UpdateViews();
                }
            });
        }

        private void ViewModelRefreshError(object sender, string s)
        {
            this.Activity.RunOnUiThread(() =>
            {
                if (this.IsVisible)
                {
                    this._swipeRefreshLayout.Refreshing = this.ViewModel.IsDataRefreshing;
                    Snackbar.Make(this._swipeRefreshLayout, s, Snackbar.LengthLong).Show();
                }
            });
        }

        private void ViewModelRefreshSuccess(object sender, string s)
        {
            this.Activity.RunOnUiThread(() =>
            {
                if (this.IsVisible)
                {
                    this._swipeRefreshLayout.Refreshing = this.ViewModel.IsDataRefreshing;
                    Snackbar.Make(this._swipeRefreshLayout, s, Snackbar.LengthLong).Show();
                }
            });
        }
    }
}